controller/cmd/service-mirror/probe_worker.go

```diff
@@ -0,0 +1,140 @@
+package servicemirror
+
+import (
+	"fmt"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	logging "github.com/sirupsen/logrus"
+)
+
+const httpGatewayTimeoutMillis = 50000
+
+// ProbeWorker is responsible for monitoring gateways using a probe specification
+type ProbeWorker struct {
+	*sync.RWMutex
+	probe          *GatewayProbeSpec
+	pairedServices map[string]bool
+	stopCh         chan struct{}
+	metrics        *probeMetrics
+	log            *logging.Entry
+}
+
+// NewProbeWorker creates a new probe worker associated with a particular gateway
+func NewProbeWorker(probe *GatewayProbeSpec, metrics *probeMetrics, probekey string) *ProbeWorker {
+	return &ProbeWorker{
+		RWMutex:        &sync.RWMutex{},
+		probe:          probe,
+		pairedServices: make(map[string]bool),
```

Comment:
I like using `map[string]struct{}` for stuff like this.  The problem with storing a boolean value is that it's easy to make the mistake of setting the value to false when unpairing a service (rather than deleting the key) which would cause a bug.